### PR TITLE
Catalog storage and Fetch on create/update broker

### PR DESCRIPTION
**Pull request by KirilKabakchiev**
The pull request extends the SM DB to store broker catalog information. This information is needed for the aggregated catalog API as well as for the enable-service-access CF proxy flow. The PR also adds a query parameter to the get brokers API to allow returning the catalog.

- [x] Initial implementation
- [x] Refactoring
- [x] Unit tests
- [x] godoc/linter
- [x] Ready for Review

### CODE REVIEW

File: api/broker/broker_controller.go [hunk 1]
```diff
@@ -85,94 +97,144 @@ func (ctrl *Controller) addBroker(response http.ResponseWriter, request *http.Re
...
+	err = common.HandleNotFoundError(err, "broker", brokerID)
+	if err != nil {
+		return err
+	}
+
+	updateData, err := json.Marshal(updateBroker)
+	if err != nil {
+		return err
+	}
+	if err := json.Unmarshal(updateData, broker); err != nil {
```
Comment (georgifarashev) [inline api/broker/broker_controller.go]: There seems to be some problem with the Unmarshaling and timestamps (I think?) The "created_at" which is returned as a response is always "0001-01-01T00:00:00Z". The value in the DB is as expected. Maybe the whole marshal/unmarshal could be removed.

File: api/broker/broker_controller.go [hunk 2]
```diff
@@ -85,94 +97,144 @@ func (ctrl *Controller) addBroker(response http.ResponseWriter, request *http.Re
...
+func (c *Controller) getAllBrokers(response http.ResponseWriter, request *http.Request) error {
 	logrus.Debug("Getting all brokers")
-
-	brokers, err := ctrl.BrokerStorage.GetAll()
+	brokers, err := c.BrokerStorage.GetAll()
 	if err != nil {
 		return err
 	}
+	withCatalog := request.FormValue(catalogParam)
+	if strings.ToLower(withCatalog) != "true" {
```
[the remark this answers is not in the document]
Comment (KirilKabakchiev) [reply api/broker/broker_controller.go]: yep we would need to rework the storage abit so that you can specify which columns  of the table you need .. This way we won't have to do stuff like credentials = nil and catalog = nil. 
However, I am not sure if we can easily rewrite the db layer this way.

Another option is to have a db layer method getBrokerWithCatalog (but I don't like the approach)   because then we also need GetBrokerWithCatalog, etc etc.. - it becomes too hardcoded.

Basically, I went for the simplest way here - I am open for suggestions, though.

File: api/broker/broker_controller.go [hunk 3]
```diff
@@ -85,94 +97,144 @@ func (ctrl *Controller) addBroker(response http.ResponseWriter, request *http.Re
...
+	if err := json.Unmarshal(updateData, broker); err != nil {
+		return err
+	}
+
+	catalog, err := c.getBrokerCatalog(broker)
+	if err != nil {
+		return types.NewErrorResponse(err, http.StatusBadRequest, "BadRequest")
+	}
+
+	isCatalogModified := !bytes.Equal(broker.Catalog, catalog)
```
[the remark this answers is not in the document]
Comment (KirilKabakchiev) [reply api/broker/broker_controller.go]: I thought it would be faster this way (basically if its not modifed we don't provided it to the db query) -I havent benchmarked it though - it  would work without this check as well, too

File: api/broker/broker_controller.go (hunk 1 above)
Comment (KirilKabakchiev) [reply api/broker/broker_controller.go]: I am not sure that it can be removed - we need some way to get the whole updated broker. 
There are two reasons we need the updated broker 
1. As per specification we want to return it in the response (this is easily changeable and we may return empty response too). (Basically we need to merge the updated broker with the broker that we already obtained from db)
2. let's say the request contains only a new updated brokerURL - getBrokerCatalog needs more info than just URL. So we need a smart way to merge the broker from db with the request object. 

I will check what's up with the created field.

File: api/broker/broker_controller.go [hunk 4]
```diff
@@ -85,94 +97,144 @@ func (ctrl *Controller) addBroker(response http.ResponseWriter, request *http.Re
 	broker.CreatedAt = currentTime
 	broker.UpdatedAt = currentTime
 
-	err = ctrl.BrokerStorage.Create(broker)
+	catalog, err := c.getBrokerCatalog(broker)
+	if err != nil {
+		return err
```
[the remark this answers is not in the document]
Comment (KirilKabakchiev) [reply api/broker/broker_controller.go]: Isnt it always 500 for the client if we couldnt fetch the catalog ? Fetching the catalog is "Internal server error" for our API and is not something we should expose to the caller that is trying to create/ update a broker (I think?)

File: api/broker/broker_controller.go (hunk 1 above)
Comment (pankrator) [reply api/broker/broker_controller.go]: Why don't we first update the broker in the database, then do a get broker from the database, they will be already merged. Then with this information we can do catalog fetch and return the updated broker to the client

Comment (KirilKabakchiev) [reply api/broker/broker_controller.go]: If the catalog fetch fails we would need to do delete broker from the database. It would be better if we can avoid it (i.e. save in db only if fetching of catalog was successful).

File: api/api.go [hunk 5]
```diff
@@ -33,7 +34,8 @@ func Default(storage storage.Storage, env server.Environment) rest.API {
 	return &smAPI{
 		controllers: []rest.Controller{
 			&broker.Controller{
-				BrokerStorage: storage.Broker(),
+				BrokerStorage:       storage.Broker(),
```
[the remark this answers is not in the document]
Comment (KirilKabakchiev) [reply api/api.go]: In the pflags PR I've redisgned this a bit so it might not be necessary to create constructors for the controllers at all (yet)
